File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays){ 
        Map<String, Integer> genrePlayCount = new HashMap<>();
        Map<String, ArrayList<Song>> genreSongs = new HashMap<>();
        
        for(int i = 0; i < genres.length; i++){
            String genre = genres[i];
            int play = plays[i];
            
            genrePlayCount.put(genre, genrePlayCount.getOrDefault(genre, 0) + play);

            
            if(!genreSongs.containsKey(genre)){
                genreSongs.put(genre, new ArrayList<>());
            }
            genreSongs.get(genre).add(new Song(i, play));
        }
        
        List<String> sortedGenres = new ArrayList<>(genrePlayCount.keySet());
        sortedGenres.sort((g1, g2) -> genrePlayCount.get(g2) - genrePlayCount.get(g1));
        
        ArrayList<Integer> answer = new ArrayList<>();
        
        for(String genre: sortedGenres){
            ArrayList<Song> songs = genreSongs.get(genre);
            
            Collections.sort(songs);
            
            answer.add(songs.get(0).id);
            if(songs.size() > 1){
                answer.add(songs.get(1). id);
            }
        }
        
        return answer.stream().mapToInt(i->i).toArray();
    }
    
    class Song implements Comparable<Song>{
        int id;
        int plays;
        
        public Song(int id, int plays){
            this.id = id;
            this.plays = plays;
        }
        
        @Override
        public int compareTo(Song other){
            if(this.plays != other.plays){
                return other.plays - this.plays;
            }
            return this.id - other.id;
        }
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    dicts = {}
    n = len(genres)
    
    for i in range(n):
        genre = genres[i]
        
        if genre not in dicts:
            dicts[genre] = [[plays[i], -1]]
        else:
            dicts[genre][0][0] += plays[i]
        dicts[genre].append([plays[i], i])
    
    # 속한 노래가 많이 재생된 장르 순서대로 정렬
    sorted_dicts = dict(sorted(dicts.items(), key=lambda x: x[1][0][0], reverse=True))
    
    for genre_type in sorted_dicts.values():
        sorted_genre_type = sorted(genre_type, key= lambda x : (x[0], -x[1]))
        
        answer.append(sorted_genre_type[-2][1])
        if(len(sorted_genre_type) >= 3):
            answer.append(sorted_genre_type[-3][1])
        
        
    return answer